### PR TITLE
ESQL: Support ST_CONTAINS and ST_WITHIN

### LINKED ISSUE #104874: Support ST_INTERSECTS in ES|QL
### Description

This is a path-finder feature that could open the door to a large number of similar features.

```
FROM airports
| EVAL polygon = TO_GEOSHAPE("POLYGON((...))")
| WHERE ST_INTERSECTS(location, polygon)
```

See the PostGIS documentation for inspiration: https://postgis.net/docs/ST_Intersects.html

The feature can be developed in two phases:

* Bring back from prototype from last summer
  * Expects constant string WKT
  * Does in-memory constant->triangle-tree and then internal intersects
* LocalPhysicalPlanner support for push-down to Lucene
  * Lucene has native support for this, so we can optimize to use that

```[tasklist]
### Tasks
- [x] Bring back prototype from last summer (geo_point in polygon)
- [x] Support cartesian point in polygon
- [x] Support polyon intersects polygon
- [ ] Support any geometry intersects any geometry
- [ ] Support doc-values for point in polygon
- [ ] Support doc-values for geometry intersects geometry
```

### PR DESCRIPTION
**Pull request by craigtaverner**
This was a relatively small update on top of the main work on ST_INTERSECTS at https://github.com/elastic/elasticsearch/pull/104907. Most of the code is either test code, or generated code, since the foundations in `SpatialRelatesFunction` and related classes covered most of the needs.

We do both `ST_CONTAINS` and `ST_WITHIN` at the same time, due to the commutative relationship between these two. In particular we want to move literals to the right field, to both minimise the number of evaluators and simplify Lucene push-down. This is only possible if we can also swap `ST_CONTAINS` with `ST_WITHIN`, since:

```
  ST_CONTAINS(a, b) = ST_WITHIN(b, a)
```

Tasks:

* [x] Implement SpatialContains, and add it to PlanNamedTypes and EsqlFunctionRegistry
* [x] Implement SpatialContainsTests and make all tests pass:
  * [x] Fix multi-components (See [LatLongShape.makeContainsGeometryQuery](https://github.com/apache/lucene/blob/c78533c53cba8072cecabf62269ec8f0bfb35bed/lucene/core/src/java/org/apache/lucene/document/LatLonShape.java#L355) and [ShapeDocValuesQuery.getContainsWeight](https://github.com/elastic/elasticsearch/blob/cf5fbfc857b36438baf357a31a7c2bc960111797/server/src/main/java/org/elasticsearch/lucene/spatial/ShapeDocValuesQuery.java#L165)
  * [x] Fix rectangles crossing dateline (see Lucene's Rectangle2D)
* [x] Implement `ST_WITHIN` to avoid needing too many evaluators (support inverting between contains and within)
  * [x] Implement SpatialWithin, and add it to PlanNamedTypes and EsqlFunctionRegistry
  * [x] Implement LogicalPlanOptimizer rule to invert between contains and within when left argument is constant and right is not
  * [x] Implement SpatialWithinTests and make all tests pass:
* [x] Implement integration tests (using existing airports and airport_city_boundaries datasets)
  * [x] Basic tests loading data from airports and city_boundaries and performing predicates
  * [x] Tests together with ST_CENTROID for doc-values support
  * [x] tests with EVAL with and without centroid
* [x] Develop Lucene-pushdown in `LocalPhysicalPlanOptimizer` and `PhysicalPlanOptimizerTests`
  * [x] Validate interaction with point doc-values (`ST_CONTAINS` and `ST_WITHIN` before `ST_CENTROID`)
* [x] Write documentation
* [x] Fix contains/within for Cartesian (currently disabled)
  * [x] Add more comprehensive multi-polygon integration tests for cartesian intersects/contains/within
  * [x] Re-enable lucene push-down for contains/within with cartesian data
  * [x] Fix any remaining issues discovered from these integration tests in lucene push-down
* [ ] Either implement `ST_DISJOINT` or create an issue to do that

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 1]
```diff
@@ -175,6 +177,45 @@ protected static List<Batch<LogicalPlan>> rules() {
...
+            return plan;
+        }
+
+        private SpatialRelatesFunction invertIfNecessary(SpatialRelatesFunction srf) {
+            if (srf.isCommutative()) {
+                // Never bother to invert commutative functions
+                return srf;
+            }
+            if (srf.left().foldable() && srf.right().foldable() == false) {
+                // If we have a constant on the left, and a field on the right, invert the function
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:212]: 1. Why is this necessary/helpful?
2. Is this really equivalent in all cases? It looks to me like inverting e.g. `SpatialWithin` or `SpatialContains` produces a non-equivalent condition. Don't we need to negate the whole thing for this to work?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialRelatesFunction.java [hunk 2]
```diff
@@ -203,13 +203,17 @@ public boolean rightDocValues() {
     /**
      * Produce a map of rules defining combinations of incoming types to the evaluator factory that should be used.
      */
-    protected abstract Map<SpatialEvaluatorFactory.SpatialEvaluatorKey, SpatialEvaluatorFactory<?, ?>> evaluatorRules();
+    abstract Map<SpatialEvaluatorFactory.SpatialEvaluatorKey, SpatialEvaluatorFactory<?, ?>> evaluatorRules();
+
+    public abstract boolean isCommutative();
+
+    public abstract SpatialRelatesFunction invert();
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialRelatesFunction.java:210]: I don't think we need this - `BinaryScalarFunction` already has `replaceChildren` which should do what you need.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 3]
```diff
@@ -175,6 +177,45 @@ protected static List<Batch<LogicalPlan>> rules() {
...
+                return srf.invert();
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:213]: This looks very close to `LiteralsOnTheRight`. Is there a chance we can just use that one? E.g. by having `SpatialRelatesFunction` inherit from `BinaryOperator` rather than from `BinaryScalarFunction` (that may be a big ask, though, not sure that would work); or by generalizing `LiteralsOnTheRight` to handle both cases?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialRelatesFunction.java (hunk 2 above)
Comment (alex-spies) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialRelatesFunction.java:210]: Woops, never mind - `replaceChildren` won't let you exchange the left and right doc values, sorry!

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java (hunk 1 above)
Comment (alex-spies) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:212]: Scratch No.2 - I just realized that within/contains gets replaced by the other, resp.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java (hunk 3 above)
Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:213]: I did investigate BinaryOperator for the original SpatialIntersects function, but it turns out BinaryOperator has a lot of assumptions that are not applicable here, so I backed out of that approach. I'll take a peek at LiteralsOnTheRight, to see if it is amenable to this.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 4]
```diff
@@ -177,42 +177,15 @@ protected static List<Batch<LogicalPlan>> rules() {
...
-                // Never bother to invert commutative functions
-                return srf;
-            }
-            if (srf.left().foldable() && srf.right().foldable() == false) {
-                // If we have a constant on the left, and a field on the right, invert the function
-                return srf.invert();
-            }
-            return srf;
+        protected SpatialRelatesFunction rule(SpatialRelatesFunction function) {
+            return function.surrogate();
```
Comment (alex-spies) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:188]: Looks good in the new version!

Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:188]: Thanks, and I am adding a comment explaining the similarity to the two other approaches, and the decision to go with this approach for now.

File: docs/reference/esql/functions/st_contains.asciidoc [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+*Example*
+
+[source.merge.styled,esql]
+----
+include::{esql-specs}/spatial.csv-spec[tag=st_contains-airport_city_boundaries]
+----
+[%header.monospaced.styled,format=dsv,separator=|]
+|===
+include::{esql-specs}/spatial.csv-spec[tag=st_contains-airport_city_boundaries-results]
+|===
```
Comment (nik9000) [inline docs/reference/esql/functions/st_contains.asciidoc:34]: You can *generally* just link to `esql/functions/layout/st_contains.asciidoc` - especially if you add an `@Example` to `@FunctionInfo` annotation. Then it *should*  spit out pretty much the same thing as this. You may want a `note` for the "The second parameter must al..." part.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialContains.java [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+            }
+            return true;
+        }
+    }
+
+    @FunctionInfo(returnType = { "boolean" }, description = "Returns whether the first geometry contains the second geometry.")
+    public SpatialContains(
+        Source source,
+        @Param(
+            name = "geomA",
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialContains.java:102]: I'd use a full word. This will be used by Kibana for autocomplete, but I am not sure if as is or they will give it a proper name.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialWithin.java [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+    @FunctionInfo(returnType = { "boolean" }, description = "Returns whether the first geometry is within the second geometry.")
+    public SpatialWithin(
+        Source source,
+        @Param(
+            name = "geomA",
+            type = { "geo_point", "cartesian_point", "geo_shape", "cartesian_shape" },
+            description = "Geometry column name or variable of geometry type"
+        ) Expression left,
+        @Param(
+            name = "geomB",
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialWithin.java:75]: Same here about the parameter names.

**Review by astefan — APPROVED** [review]
Two minor comments. LGTM

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/LuceneComponent2DUtils.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+ * This utilities class provides access to protected methods in Lucene using alternative APIs.
+ * For example, the 'create' method returns the original Component2D array, instead of a Component2D containing
+ * a component tree of potentially multiple components. This is particularly useful for algorithms that need to
+ * operate on each component individually.
+ */
+public class LuceneComponent2DUtils {
+    /**
+     * This method is based on LatLonGeometry.create, but returns an array of Component2D objects for multi-component geometries.
+     */
+    public static Component2D[] createLatLonComponents(LatLonGeometry... latLonGeometries) {
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/LuceneComponent2DUtils.java:28]: Move towards InvalidArgumentException since those have proper HTTP codes.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java [hunk 9]
```diff
@@ -1485,6 +1489,28 @@ static void writeIntersects(PlanStreamOutput out, SpatialIntersects intersects)
...
 
+    static SpatialContains readContains(PlanStreamInput in) throws IOException {
+        return new SpatialContains(Source.EMPTY, in.readExpression(), in.readExpression());
+    }
+
+    static void writeContains(PlanStreamOutput out, SpatialContains intersects) throws IOException {
+        List<Expression> fields = intersects.children();
+        assert fields.size() == 2;
+        out.writeExpression(fields.get(0));
+        out.writeExpression(fields.get(1));
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java:1500]:
```
out.writeExpression(intersects.left())
out.writeExpression(intersects.right())
```
No need to do assertion or generic extractions.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java [hunk 10]
```diff
@@ -1485,6 +1489,28 @@ static void writeIntersects(PlanStreamOutput out, SpatialIntersects intersects)
...
+    static SpatialWithin readWithin(PlanStreamInput in) throws IOException {
+        return new SpatialWithin(Source.EMPTY, in.readExpression(), in.readExpression());
+    }
+
+    static void writeWithin(PlanStreamOutput out, SpatialWithin intersects) throws IOException {
+        List<Expression> fields = intersects.children();
+        assert fields.size() == 2;
+        out.writeExpression(fields.get(0));
+        out.writeExpression(fields.get(1));
+    }
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java:1512]: Same as above.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java [hunk 11]
```diff
@@ -174,6 +176,25 @@ protected static List<Batch<LogicalPlan>> rules() {
...
     }
 
+    /**
+     * Currently this works similarly to SurrogateExpression, leaving the logic inside the expressions,
+     * so each can decide for itself whether or not to change to a surrogate expression.
+     * But what is actually being done is similar to LiteralsOnTheRight. We can consider in the future moving
+     * this in either direction, reducing the number of rules, but for now,
+     * it's a separate rule to reduce the risk of unintended interactions with other rules.
+     */
+    static class SubstituteSpatialSurrogates extends OptimizerRules.OptimizerExpressionRule<SpatialRelatesFunction> {
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:186]: Move it under SubstituteSurrogate and please raise an issue to indicate that this needs to either be updated based on either LiteralsOnTheRight OR be incorporated into SubstituteSurrogates

**Review by costin — APPROVED** [review]
LGTM

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialContains.java (hunk 6 above)
Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialContains.java:103]: I used `geomA` and `geomB` because that is exactly what PostGIS used, and I kind-of liked the idea that our docs look a little similar to PostGIS docs. But Kibana obviously takes precedence. I could ask them about this.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/LuceneComponent2DUtils.java (hunk 8 above)
Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/LuceneComponent2DUtils.java:28]: Right now, all the spatial functions use IllegalArgumentException, and that was based on copying from other functions. I think I should move the whole lot over to InvalidArgumentException in one go, in a separate PR, if that is OK for you?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java (hunk 9 above)
Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java:1500]: Done, and merged the three write functions into one, since they are identical. But we still need three read functions to instantiate the correct classes.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java (hunk 11 above)
Comment (craigtaverner) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:187]: Moved and created issue at https://github.com/elastic/elasticsearch/issues/106763

File: docs/reference/esql/functions/st_contains.asciidoc (hunk 5 above)
Comment (craigtaverner) [reply docs/reference/esql/functions/st_contains.asciidoc:34]: Cool! I'm trying the @Example option, but will leave the rest of the options alone in this PR and play with this more in later work.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/querydsl/query/SpatialRelatesQuery.java [hunk 12]
```diff
@@ -231,9 +231,9 @@ private static org.apache.lucene.search.Query pointShapeQuery(
...
             final boolean hasDocValues = context.getFieldType(fieldName).hasDocValues();
-            // only the intersects relation is supported for indexed cartesian point types
-            if (relation != ShapeField.QueryRelation.INTERSECTS) {
-                throw new QueryShardException(context, relation + " query relation not supported for Field [" + fieldName + "].");
+            if (geometry.type() != ShapeType.POINT && relation == ShapeField.QueryRelation.CONTAINS) {
+                // A point field can never contain a non-point geometry
+                return new MatchNoDocsQuery();
             }
             final Consumer<ShapeType> checker = t -> {
                 if (t == ShapeType.POINT || t == ShapeType.MULTIPOINT || t == ShapeType.LINESTRING || t == ShapeType.MULTILINESTRING) {
```
Comment (iverase) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/querydsl/query/SpatialRelatesQuery.java:239]: I wonder if this checker is correct here. I think in cartesian we can generate any geometry because there is no loosy quantization so everything is in theory supported.
